rig/machine_control/machine_controller.py

```diff
@@ -1739,34 +1807,169 @@ def get_machine(self, x=0, y=0, default_num_cores=18):
+                                 sdram_resource=SDRAM,
+                                 sram_resource=SRAM,
+                                 _all_chips_info=None):
+        """Returns a set of place-and-route
+        :py:class:`~rig.place_and_route.constraints` which reserves any cores
+        and SDRAM that are already in use.
+
+        .. note::
+            Most users will probably prefer to use
+            :py:meth:`.get_machine_and_resource_constraints` which both gets a
+            :py:class:`rig.machine.Machine` object and a set of place-and-route
+            constraints with reduced overhead over calling both methods
+            individually.
+
+        This method, like :py:meth:`get_machine`, is aimed at users of Rig's
+        place-and-route tools and assumes that one resource type represents
+        Cores while another represents SDRAM.
+
+        The returned set of
+        :py:class:`~rig.place_and_route.constraints.ReserveResourceConstraint`s
+        reserves all cores not in an Idle state (i.e. not a monitor and not
+        already running an application).
+
+        This set also reserves ranges of SDRAM/SRAM starting at the top
+        (highest address) to reduce the resource availability on each chip to
+        approximately the amount actually available in the machine, accounting
+        for any already-running applications. Note that due to memory
+        fragmentation this constraint will make a conservative estimate based
+        on the largest free block of SDRAM/SRAM in the heap.
+
+        Parameters
+        ----------
+        machine : :py:class:`~rig.machine.Machine`
+            The :py:class:`~rig.machine.Machine` model of the current SpiNNaker
+            machine (e.g. from :py:meth:`get_machine`).
+        core_resource : resource (Default: :py:data:`~rig.machine.Cores`)
+            The resource identifier used for cores. If using
+            :py:meth:`get_machine`, the default correct.
+        sdram_resource : resource (Default: :py:data:`~rig.machine.SDRAM`)
+            The resource identifier used for SDRAM. If using
+            :py:meth:`get_machine`, the default correct.
+        sram_resource : resource (Default: :py:data:`~rig.machine.SRAM`)
+            The resource identifier used for SRAM. If using
+            :py:meth:`get_machine`, the default correct.
+        _all_chips_info : None
+            For internal use only. Users should leave this as None.
+
+        Returns
+        -------
+        [\
+        :py:class:`rig.place_and_route.constraints.ReserveResourceConstraint`,\
+        ...]
+            A set of place-and-route constraints which reserves non-idle cores
+            and SDRAM. The resource types given in the ``core_resource`` and
+            ``sdram_resource`` arguments will be reserved accordingly.
+        """
+        if _all_chips_info is not None:
+            chips_info = _all_chips_info[2]
+        else:
+            chips_info = {}
```

Comment:
The factory function to a defaultdict doesn't get given any arguments so sadly this is not possible.